staging/src/kubesphere.io/api/alerting/v2beta1/rulegroup_webhook.go

```diff
@@ -0,0 +1,241 @@
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v2beta1
+
+import (
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/rulefmt"
+	yaml "gopkg.in/yaml.v3"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+var rulegrouplog = logf.Log.WithName("rulegroup")
+
+func (r *RuleGroup) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+var _ webhook.Defaulter = &RuleGroup{}
+
+func (r *RuleGroup) Default() {
+	log := rulegrouplog.WithValues("name", r.Namespace+"/"+r.Name)
+	log.Info("default")
+	for i := range r.Spec.Rules {
+		rule := r.Spec.Rules[i]
+		if rule.ExprBuilder != nil {
+			if rule.ExprBuilder.Workload != nil {
+				rule.Expr = intstr.FromString(rule.ExprBuilder.Workload.Build())
+			}
+		}
+		r.Spec.Rules[i] = rule
+	}
+}
+
+var _ webhook.Validator = &RuleGroup{}
+
+func (r *RuleGroup) ValidateCreate() error {
+	return r.Validate()
+}
+func (r *RuleGroup) ValidateUpdate(old runtime.Object) error {
+	return r.Validate()
+}
+func (r *RuleGroup) ValidateDelete() error {
+	return nil
+}
+func (r *RuleGroup) Validate() error {
+	log := rulegrouplog.WithValues("name", r.Namespace+"/"+r.Name)
+	log.Info("validate")
+
+	var rules []Rule
+	for _, r := range r.Spec.Rules {
+		rules = append(rules, r.Rule)
+	}
+	var err = validateRules(log, r.Name, r.Spec.Interval, rules)
+	if err == errorEmptyExpr {
+		return fmt.Errorf("one of 'expr' and 'exprBuilder.workload' must be set for a RuleGroup")
+	}
+	return err
+}
+
+type ruleGroup struct {
+	Name     string         `yaml:"name"`
+	Interval model.Duration `yaml:"interval,omitempty"`
+	Rules    []rulefmt.Rule `yaml:"rules"`
+}
+
+type ruleGroups struct {
+	Groups []ruleGroup `yaml:"groups"`
+}
+
+var errorEmptyExpr = fmt.Errorf("'expr' is empty")
+
+func validateRules(log logr.Logger, groupName, groupInterval string, rules []Rule) error {
+	durationStr := groupInterval
+	if durationStr == "" {
+		durationStr = "1m"
+	}
+	interval, err := model.ParseDuration(durationStr)
+	if err != nil {
+		return fmt.Errorf("invalid 'interval': %s", durationStr)
+	}
+
+	var g = ruleGroup{
+		Name:     groupName,
+		Interval: interval,
+	}
+
+	for i := range rules {
+		rule := rules[i]
+		if rule.Alert == "" {
+			return fmt.Errorf("'alert' cannot be empty")
+		}
+		if rule.Expr.String() == "" {
```

Comment:
No Prometheus check for expr?

Reply:
The following codes will call the prometheus libs to verify it: 
```go
_, errs := rulefmt.Parse(content)
```